python/ray/ml/train/integrations/sklearn/sklearn_trainer.py

```diff
@@ -112,26 +152,282 @@ def __init__(
+        parallelize_cv = False
+
+        assert not (has_gpus and self.parallelize_cv)
+
+        if self.cv and self.parallelize_cv is True:
+            estimator_has_parallelism_params = self._set_estimator_cpus(
+                self.estimator, 1
+            )
+            parallelize_cv = True
+        else:
+            estimator_has_parallelism_params = self._set_estimator_cpus(
+                self.estimator, num_cpus
+            )
+
+        if (
+            not has_gpus
+            and self.cv
+            and self.parallelize_cv is None
+            and not estimator_has_parallelism_params
+        ):
+            logger.info(
+                "No parallelism-related params detected in estimator, "
+                "will parallelize cross-validation instead."
+            )
+            parallelize_cv = True
+        return parallelize_cv
 
     def training_loop(self) -> None:
-        return super().training_loop()
+        register_ray()
+
+        self.estimator.set_params(**self.params)
+
+        datasets = self._get_datasets()
+        X_train, y_train = datasets.pop(TRAIN_DATASET_KEY)
+        groups = None
+        if "cv_groups" in X_train.columns:
+            groups = X_train["cv_groups"]
+            X_train = X_train.drop("cv_groups", axis=1)
+
+        trial_resources = tune.get_trial_resources().required_resources
+        num_cpus = int(trial_resources.get("CPU", 0))
+        has_gpus = bool(trial_resources.get("GPU", 0))
+
+        # see https://scikit-learn.org/stable/computing/parallelism.html
+        os.environ["OMP_NUM_THREADS"] = str(num_cpus)
+        os.environ["MKL_NUM_THREADS"] = str(num_cpus)
+        os.environ["OPENBLAS_NUM_THREADS"] = str(num_cpus)
+        os.environ["BLIS_NUM_THREADS"] = str(num_cpus)
+
+        parallelize_cv = self._set_estimator_cv_parallelism(num_cpus, has_gpus)
+
+        with parallel_backend("ray", n_jobs=num_cpus):
+            start_time = time()
+            self.estimator.fit(X_train, y_train, **self.fit_params)
+            fit_time = time() - start_time
+
+            with tune.checkpoint_dir(step=1) as checkpoint_dir:
+                with open(os.path.join(checkpoint_dir, MODEL_KEY), "wb") as f:
+                    cpickle.dump(self.estimator, f)
+
+            if self.label_column:
+                validation_set_scores = self._score_on_validation_sets(
+                    self.estimator, datasets
+                )
+                cv_scores = self._score_cv(
+                    self.estimator,
+                    X_train,
+                    y_train,
+                    groups,
+                    # if estimator has parallelism, use that. Otherwise,
+                    # parallelize CV
+                    n_jobs=1 if not parallelize_cv else num_cpus,
+                )
+            else:
+                validation_set_scores = {}
+                cv_scores = {}
+
+        results = {
+            **validation_set_scores,
+            **cv_scores,
```

Comment:
keys from validation_set_scores and cv_scores will never override each other?

Reply:
Never

Reply:
done